### PR TITLE
Support focus stealing filters

### PR DESCRIPTION
**Pull request by Elv13**
After seeing the "Fixing Firefox focus stealing" mailing list thread.

the goal of this is "simply to make it simple".

### CODE REVIEW

File: lib/awful/ewmh.lua [hunk 1]
```diff
@@ -68,6 +85,55 @@ function ewmh.activate(c, context, hints) -- luacheck: no unused args
...
+-- The callback takes the following parameters:
+--
+-- * **c** (*client*) The client requesting the activation
+-- * **context** (*string*) The activation context.
+-- * **hints** (*table*) Some additional hints (depending on the context)
+--
+-- If the callback returns `true`, the client will be activated unless the `force`
+-- hint is set. If the callback returns `false`, the activation request is
+-- cancelled. If the callback returns `nil`, the previous callback will be
+-- executed. This will continue until either a callback handle the request or
```
Comment (blueyed) [inline lib/awful/ewmh.lua]: s/handle/handles/

File: lib/awful/ewmh.lua [hunk 2]
```diff
@@ -68,6 +85,55 @@ function ewmh.activate(c, context, hints) -- luacheck: no unused args
...
+-- when it run out of callbacks. In that case, the request will be granted if
```
Comment (blueyed) [inline lib/awful/ewmh.lua]: s/run/runs/

File: lib/awful/ewmh.lua [hunk 3]
```diff
@@ -68,6 +85,55 @@ function ewmh.activate(c, context, hints) -- luacheck: no unused args
...
+--    end)
+--
+-- @tparam function f The callback
+-- @see activate_filters
+-- @see remove_activate_filter
+function ewmh.add_activate_filter(f)
+    table.insert(ewmh.activate_filters, f)
+end
+
+--- Remove an activate (focus steanling) filter function.
```
Comment (blueyed) [inline lib/awful/ewmh.lua]: s/steanling/stealing/

File: tests/test-awful-client.lua [hunk 4]
```diff
@@ -77,6 +77,41 @@ end
...
+
+    -- This should still be the case
+    assert(client.focus == c1)
+
+    c2:emit_signal("request::activate", "i_said_so")
+
+    return true
+end)
+
+-- Check if writing a focus stealing filter work
```
Comment (blueyed) [inline tests/test-awful-client.lua]: s/work/works./

**Comment by Elv13** [comment]
Done

> I am using some advanced stuff in this regard myself, where I wrap firefox to write a hint to a file, so that it gets whitelisted/raised when executed explicitly (e.g. via xdg-open). This will help me in that regard.

At some point you should share your config... I think @mindeunix uses an SQLite database for this.

Personally, I wrap this into awesome-client "awful.spawn(cmd, {rules})". No need to serialize any hints. They are pushed directly into Awesome.

**Comment by Elv13** [comment]
I added an unrelated commit to fix the tests. Maybe `luacheck` was updated? The warning is "real", so it has to be fixed anyway.

File: lib/awful/ewmh.lua [hunk 5]
```diff
@@ -68,6 +85,55 @@ function ewmh.activate(c, context, hints) -- luacheck: no unused args
...
+-- @treturn boolean If the callback existed
+-- @see activate_filters
+-- @see add_activate_filter
+function ewmh.remove_activate_filter(f)
+    for k, v in ipairs(ewmh.activate_filters) do
+        if v == f then
+            table.remove(ewmh.activate_filters, k)
+
+            -- In case the callback is there multiple time.
+            ewmh.remove_activate_filter(f)
```
Comment (psychon) [inline lib/awful/ewmh.lua]: Is the order of these filters really relevant? I'd expect them to be side-effect-free aka "their order is relevant". If so, I'd simplify the filter management:
```lua
function add(f)
  ewmh.activate_filters[t] = true
end
function remove(f)
  local old = ewmh.activate_filters[t]
  ewmh.activate_filters[t] = nil
  return old == true
end
```

Comment (Elv13) [inline lib/awful/ewmh.lua]: > Is the order of these filters really relevant?

Yes, just like other rules. It goes from the specific to the generic

File: tests/test-awful-client.lua [hunk 6]
```diff
@@ -77,6 +77,61 @@ end
...
+
+    return true
+end)
+
+
+table.insert(steps, function()
+    -- Now that the filter is gone, it should work again.
+
+    return client.focus == c1
+end)
```
Comment (psychon) [inline tests/test-awful-client.lua]: I think that you can merge all of this into a single step. While it is not always clear which operations are instantaneous and which aren't, setting the input focus works fine (except for some corner cases where the client then forwards the focus elsewhere).
Less steps = faster tests. :-)

File: .luacheckrc [hunk 7]
```diff
@@ -23,13 +23,14 @@ read_globals = {
...
     "root",
-    "screen",
     "selection",
     "tag",
     "window",
 }
+
 -- May not be read-only due to client.focus
 globals = {
+    "screen",
```
Comment (psychon) [inline .luacheckrc]: Could add to the commit message and the comment before this array the reason why `screen` can not be read-only?

**Comment by psychon** [comment]
> I think @mindeunix uses an SQLite database for this.

Wtf? Are there usable SQLite-Lua bindings? And... wtf?

> Personally, I wrap this into awesome-client "awful.spawn(cmd, {rules})". No need to serialize any hints. They are pushed directly into Awesome.

Just to be sure: You use the above to add a callback and that callback then adds a "may no longer steal focus" entry? How exactly does that code look like?

> I added an unrelated commit to fix the tests. Maybe luacheck was updated? The warning is "real", so it has to be fixed anyway.

I'd be curious to know which warning that is exactly :-)

File: .luacheckrc (hunk 7 above)
Comment (Elv13) [inline .luacheckrc]: Well, because the new version of luacheck complain if you do `screen[1].foo = "bar"` while `local s = screen[1]; s.foo = "bar"` still work. I will add it to the commit message.

**Comment by Elv13** [comment]
> Wtf? Are there usable SQLite-Lua bindings? And... wtf?
https://github.com/mindeunix/clientdb

**Comment by Elv13** [comment]
> > Personally, I wrap this into awesome-client "awful.spawn(cmd, {rules})". No need to serialize any 
> hints. They are pushed directly into Awesome.
> 
> Just to be sure: You use the above to add a callback and that callback then adds a "may no longer
> steal focus" entry? How exactly does that code look like?

The code above is about complete. Try it for yourself

    awesome-client  "require('awful.spawn')('urxvt', {new_tag=true})"

Tyrannical has some focus stealing code of its own based on monkey-patched hacks and luck. The 3.6 version will use request::activate to implement it using this PR and the SNID based spawn.

**Comment by madduck** [comment]
I hope I am not thick, but I can't find @Elv13's branch for this…?

**Comment by psychon** [comment]
@madduck GitHub tells me I can merge it via `git pull https://github.com/Elv13/awesome-1.git focus_stealing_filter`. Does that info help you?

**Comment by madduck** [comment]
@psychon Indeed. I was looking at `awesome.git`, not `awesome-1.git`. Weird how this information doesn't seem to be apparent from the pull request.

**Comment by madduck** [comment]
Applied this on top of 7f74c783cd174edb241fcfa2a695cdc78927982d and it seems to work nicely!

**Comment by madduck** [comment]
Actually, @Elv13, there's a bug: with this applied, I can no longer use `mod4-j/k` to activate the Firefox window. Maybe this is because your filters do not check for `context == evmh`?

**Comment by Elv13** [comment]
> Maybe this is because your filters do not check for context == evmh? 

Right, the filter I proposed is too aggressive. I will have to improve this pull request to properly document request::activate and the different contexts (and fix the example).

**Comment by madduck** [comment]
For completeness, and since I noticed that you do make `context` available:

```
awful.ewmh.add_activate_filter(function(c, context, hints)
    if context == "ewmh" and c.class == "Firefox" then return false end
end
```

works fine.

**Comment by Elv13** [comment]
Thanks @madduck 

@blueyed Do you think we should have a second parameter to add "per context" filters? This seems like a common use case?

**Comment by blueyed** [comment]
@Elv13 
Providing `context` here makes sense to me, and I am also using it in my current `client.disconnect_signal("request::activate", awful.ewmh.activate)` overwrite.

**Comment by psychon** [comment]
I cherry-picked 2cd587d77e6d0aa8c043ee4c79295688ee91dc56 into master amended with a change adding a comment explaining why this is needed.

**Comment by Elv13** [comment]
All done and rebased
